ibis/timecontext/adjustment.py

```diff
@@ -0,0 +1,53 @@
+""" Time context adjustment algorithm
+    In an Ibis tree, time context is local for each node, and they should be
+    adjusted accordingly for some specific nodes. Those operations may
+    require extra data outside of the global time context that user defines.
+    For example, in asof_join, we need to look back extra `tolerance` daays
+    for the right table to get the data for joining. Similarly for window
+    operation with preceeding and following.
+    Algorithm to calculate context adjustment are defined in this module
+    and could be used by multiple backends.
+"""
+from ibis.expr.typing import TimeContext
+
+
+def adjust_context_asof_join(
```

Comment:
We should probably define dispatch methods for extending.

```
@singledispatch
def adjust_timecontext(op, timecontext):
      raise NotImplementedError()


@adjust_timecontext.register(AsOfJoin):
 def adjust_timecontext_asof_join(op, timecontext):
        return ...
```

Reply:
In order for this to work, I added an extra arg `clients` in `compute_time_context` dispatch, we will need to pass clients so that in these dispatch function at expr level, we could run `execute` on op args.